Fix Kobuki workspace CI arm64 cross-compile error

j3soon opened the pull request:
Fixed by updating upstream dependencies, didn't tested on real hardware though.

Fixes: https://github.com/j3soon/ros2-essentials/issues/72

j3soon commented:
@YuZhong-Chen please review the updates if you have time. In addition, it would be great if you have time to test on the real Kobuki hardware.

YuZhong-Chen commented:
No problem, I will test the code on the real Kobuki hardware and NV Orin next week.
But I was quite surprised that the problem could be solved simply by updating these upstream packages. 
Is this because the versions of `ecl` and `kobuki` related dependencies installed by `apt` in the Dockerfile are different?

> BTW, I’m curious about your method for updating these packages.
> Do you clone the latest repositories first and then use `git diff` to generate patches? Or is there another approach?

j3soon commented:
> No problem, I will test the code on the real Kobuki hardware and NV Orin next week.

If possible, also test it on an amd64 machine, since I'm not sure if the update will break things. (I didn't have access to real Kobuki hardware now...)

> But I was quite surprised that the problem could be solved simply by updating these upstream packages. Is this because the versions of `ecl` and `kobuki` related dependencies installed by `apt` in the Dockerfile are different?

I only fixed compile time error, which seems to require both of the following:
- using Ubuntu 22.04 runner (instead of ubuntu-latest that corresponds to Ubuntu 24.04) in commit fddbb794d7c85caac8b523fae62cff0c0a48aa4f
  - My bad for committing this fix directly on `main` instead of in this PR. (Since I intentionally make PRs not to trigger CI builds)
  - To prevent this from happening again, I now created an additional `dev` branch to test CIs more easily without needing to test on `main`.
- update the upstream `kobuki_core` package.

I haven't tested the workspace functionalities though.

I simply decided to update the packages first, since I observed that there are later upstream commits that attempt to fix (arm64) compilation issues:
- `kobuki_core` https://github.com/kobuki-base/kobuki_core/commit/4dea91ab404ce194e5b62a169a16657f095469a4
  ![](https://github.com/user-attachments/assets/3bcdf6e3-f86a-4091-a852-789a86fbbd59)
  ref: https://github.com/kobuki-base/kobuki_core/commits/3e0652dc4006d8345a86298ac05d82049747bc7f/
- I believe the `ecl_core` and `kobuki_ros` updates aren't necessary, but I've updated them anyhow.

I didn't investigate why these latest upstream changes may fix compilation errors. I simply updated them, and re-run the CI build, and then the compilation succeeded.

> > BTW, I’m curious about your method for updating these packages.
> > Do you clone the latest repositories first and then use `git diff` to generate patches? Or is there another approach?

I didn't think too much on this and used a more manual approach. Generating patches should also work though.

The commands I use look something like this:

```sh
cd <upstream_package_dir>
git init
git remote add origin <upstream_git_url>
git fetch --all
git reset <upstream_git_old_commit>
git status # check if we have modified this repo before
# if there are no modifications,
git pull
rm -rf .git
# it there are modifications, review the changes and make sure the update will not overwrite our previous changes
# (some how handle the merge by hand)
rm -rf .git
```

This is also why I asked everyone to include the upstream URL, branch name, and commit ID in our git commit message when copying upstream packages. For an example, these information is required for me to quickly perform this update.

j3soon commented:
Quick Update: Copying the latest packages doesn't reliably solve the CI build issue.

I re-ran the successful CI build and found it reported error. I think it's indeed due to some randomness during `colcon build`.

- Success: https://github.com/j3soon/ros2-essentials/actions/runs/15646718088/attempts/1
- Failure: https://github.com/j3soon/ros2-essentials/actions/runs/15646718088/attempts/2

I think we don't need to test it on real hardware just yet. Meanwhile, I'll continue investigate the issue. Thanks!

YuZhong-Chen commented:
I reviewed the failed log, and it is located [here](https://github.com/j3soon/ros2-essentials/actions/runs/15646718088/job/44101893447#step:12:5206). The last few logs also have the same error.

The error message is quite strange because if the same process ran successfully before, 
it shouldn’t be a dependency problem. This issue is truly puzzling!
